hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/balancer/ContainerBalancerSelectionCriteria.java

```diff
@@ -88,19 +93,25 @@ private boolean isContainerReplicatingOrDeleting(ContainerID containerID) {
    * 5. Container should be closed.
    * 6. If the {@link LegacyReplicationManager} is enabled, then the container should not be an EC container.
    * @param node DatanodeDetails for which to find candidate containers.
-   * @return NavigableSet of candidate containers that satisfy the criteria.
+   * @return Set of candidate containers that satisfy the criteria.
    */
-  public NavigableSet<ContainerID> getCandidateContainers(
+  public Set<ContainerID> getCandidateContainers(
       DatanodeDetails node, long sizeMovedAlready) {
-    NavigableSet<ContainerID> containerIDSet =
-        new TreeSet<>(orderContainersByUsedBytes().reversed());
     try {
-      containerIDSet.addAll(nodeManager.getContainers(node));
+      // Initialize containerSet for node
+      if (!setMap.containsKey(node)) {
+        addNodeToSetMap(node);
```

Comment:
can we add filtered container set to avoid re-filtering again when called ?

Reply:
The thing is it's possible that the container does not exists on the datanode anymore, because it's already moved to other other datanodes.